### PR TITLE
Tree - review from PR #39

### PR DESCRIPTION
**Pull request by TwsThomas**

### CODE REVIEW

File: python_scripts/trees.py [hunk 1]
```diff
@@ -186,9 +187,9 @@ def plot_decision_function(X, y, clf, ax=None):
 # Indeed, it is not a surprise. We saw earlier that a single feature will not
 # be able to separate all three species. However, from the previous analysis we
 # saw that by using both features we should be able to get fairly good results.
-# Considering the mechanism of the decision tree illustrated above, we should
+# Considering the spliting mechanism of the decision tree illustrated above, we should
```
Comment (lucyleeow) [inline python_scripts/trees.py]:
```suggestion
# Considering the splitting mechanism of the decision tree illustrated above, we should
```

File: python_scripts/trees.py [hunk 2]
```diff
@@ -186,9 +187,9 @@ def plot_decision_function(X, y, clf, ax=None):
...
 # repeat the partitioning on the resulting rectangles created by the first
-# split. In this regard, we expect that the second partition will be using
+# split. In this regard, we expect that the other partitions will be using
```
Comment (lucyleeow) [inline python_scripts/trees.py]:
```suggestion
# split. In this regard, we expect that the two partitions at the second level of the tree will be using
```

maybe too long?

File: python_scripts/trees.py [hunk 3]
```diff
@@ -326,13 +327,17 @@ def plot_decision_function(X, y, clf, ax=None):
 # in the partition >= 42 mm, we cannot differentiate well between Gentoo and
 # Chinstrap (while they are almost twice more Gentoo).
 #
-# We should use a statistical measure that uses the all class
+# We should use a statistical measure that uses all the class
 # probabilities, as the criterion to qualify the purity
-# of a partition. We will choose as an example the entropy criterion (also used
+# of a partition (which will be a leaf in a decision tree).
```
Comment (lucyleeow) [inline python_scripts/trees.py]: Partitions could be either leaf or split nodes? It's just the result of a split?

File: python_scripts/trees.py [hunk 4]
```diff
@@ -326,13 +327,17 @@ def plot_decision_function(X, y, clf, ax=None):
...
+# We will choose as an example the entropy criterion (also used
 # in scikit-learn) which is one of the possible classification criterion.
 #
-# The entropy is defined as: $H(X) = - \sum_{k=1}^{K} p(X_k) \log p(X_k)$
-#
+# The entropy $H$ of the data $X$ remaining in one leaf is defined as: 
+# $H(X) = - \sum_{k=1}^{K} p(X_k) \log p(X_k)$
+# where $p(X_k)$ stand for the probability (here proportions)
```
Comment (lucyleeow) [inline python_scripts/trees.py]: Maybe it might be nice for the equation to be on its own line?

Comment (lucyleeow) [reply python_scripts/trees.py]:
```suggestion
# where $p(X_k)$ stands for the probability (here proportions) 
```

File: python_scripts/trees.py [hunk 5]
```diff
@@ -342,7 +347,7 @@ def plot_decision_function(X, y, clf, ax=None):
 # each class will be equal and minimum when only samples for a single class
 # is present.
 #
-# Therefore, one searches to minimize the entropy in each partition.
+# Therefore, a good partition (i.e. a good leaf) *minimize* the entropy.
```
Comment (lucyleeow) [inline python_scripts/trees.py]: I think leaf means end? Would 'node' be right here?
```suggestion
# Therefore, a good partition (i.e. a good leaf) *minimizes* the entropy.
```

File: python_scripts/trees.py [hunk 6]
```diff
@@ -369,8 +374,9 @@ def classification_criterion(labels):
 # and Gentoo penguins.
 #
 # With entropy, we are able to assess the quality of each partition. However,
-# the ultimate goal is to evaluate the quality of the overall split and thus
-# combine the measures of entropy in each partition into a single statistic.
+# the ultimate goal is to evaluate the quality of the overall split
+# (the node in decision tree)and thus
```
Comment (lucyleeow) [inline python_scripts/trees.py]: I don't think 'the node in decision tree' is needed and it is a little confusing, are we talking about the child nodes?

File: python_scripts/trees.py [hunk 7]
```diff
@@ -379,8 +385,11 @@ def classification_criterion(labels):
 # of a split. The information gain is defined as the difference between the
 # entropy
 # before a split and the sum of the entropies of each partition,
-# normalized by the frequencies of class samples in each partition. The goal is
-# to maximize the information gain.
+# normalized by the frequencies of class samples in each partition. 
+# 
+# IG = H(X_unsplit)/N - ( H(split1)/N1 + H(split2)/N2 )
+# 
+# The goal is to maximize the information gain (i.e. minimizing the entropy of each split).
```
Comment (lucyleeow) [inline python_scripts/trees.py]:
```suggestion
# The goal is to maximize the information gain (i.e. maximize the decrease in entropy after the split).
```

File: python_scripts/trees.py [hunk 8]
```diff
@@ -632,17 +641,16 @@ def plot_regression_model(X, y, model, extrapolate=False, ax=None):
 _ = plot_regression_model(X_train, y_train, tree)
 
 # %% [markdown]
-# We see that the decision tree model does not have a priori and we do not
-# end-up
-# with a straight line to regress flipper length and body mass. The prediction
-# of a sample from the training set will give the
-# same target than this training sample. However, having different body masses
+# We see that the decision tree model does not have an a priori distribution
```
Comment (lucyleeow) [inline python_scripts/trees.py]:
```suggestion
# We see that the decision tree model does not have a priori distribution
```

File: python_scripts/trees.py [hunk 9]
```diff
@@ -632,17 +641,16 @@ def plot_regression_model(X, y, model, extrapolate=False, ax=None):
...
+# for the data and we do not end-up
+# with a straight line to regress flipper length and body mass.
+# Having different body masses
 # for a same flipper length, the tree will be predicting the mean of the
 # targets.
 #
 # So in classification setting, we saw that the predicted value was the most
 # probable value in the node of the tree. In the case of regression, the
-# predicted value corresponds to the mean of the target in the node.
+# predicted value corresponds to the mean of the target in the leaf.
```
Comment (lucyleeow) [inline python_scripts/trees.py:656]: leaf ?

Comment (TwsThomas) [reply python_scripts/trees.py:656]: In the final node - which is a leaf.

File: python_scripts/trees.py (hunk 3 above)
Comment (TwsThomas) [reply python_scripts/trees.py]: Yes right ! Partition here is a part of the full data.

Comment (TwsThomas) [reply python_scripts/trees.py]: Right, Partitions could be either leaf or split nodes

**Comment by lesteve** [comment]
@lucyleeow can you merge this one so that all the changes are in the PR on the main repo?

**Pull request merged**